src/stan/analyze/mcmc/compute_ess.hpp

```diff
@@ -0,0 +1,100 @@
+#ifndef STAN_ANALYZE_ESS_HPP
+#define STAN_ANALYZE_ESS_HPP
+
+#include <stan/math/prim/mat.hpp>
+#include <algorithm>
+#include <vector>
+
+namespace stan {
+  namespace analyze {
+
+    /**
+     * Returns the effective sample size for the specified parameter
+     * across all kept samples.
+     *
+     * See more details in Stan reference manual section "Effective
+     * Sample Size". http://mc-stan.org/users/documentation
+     *
+     * Current implementation assumes chains are all of equal size and
+     * draws are stored in contiguous blocks of memory.
+     *
+     * @param std::vector stores pointers to arrays of chains
+     * @param std::vector stores sizes of chains
+     * @return effective sample size for the specified parameter
+     */
+    template<typename T>
```

Comment:
Thanks for asking, as I was hoping for feedback on this.  I addressed my reasoning behind templating this function in a comment immediately following this PR's initial comment.  I've come to learn that you'd prefer such details/requests for feedback in the PR's initial comment.

The short story is that CmdStan won't have to make any copies if it can use `compute_effective_sample_size(std::vector<const double*>, std::vector<size_t> sizes)`, but everyone agreed to `compute_effective_sample_size(std::vector<double*>, std::vector<size_t> sizes)`, for which CmdStan will necessitate a copy.  In an attempt to satisfy all requests of our previous discussions, I templated this function.  

If RStan and PyStan can both call either signature, then I'd recommend we change the type of the first argument from `std::vector<double*>` to `std::vector<const double*>` and drop the template. 

What do you think?

Reply:
Thanks for bringing this up! We should use `const double *`. I think it should actually be `const double * const`. The pointers don't change and we never mutate the elements. 

CmdStan should be able to handle it. We'll figure out how to make RStan and PyStan work with that signature, even if it means casting the consts away.

Reply:
Thanks for bringing this up! We should use `const double *`. I think it should actually be `const double * const`. The pointers don't change and we never mutate the elements. 

CmdStan should be able to handle it. We'll figure out how to make RStan and PyStan work with that signature, even if it means casting the consts away.

Reply:
The `const double * const` is a good thought and I agree with the reasoning behind it, but I don't know how to deal with that type here.   Wouldn't this necessitate initialization at declaration, since it won't allow modification later?  For instance, I don't see a reasonable way get the rvalues of line 616 in src/stan/mcmc/chains.hpp into the (would be) container `std::vector<const double * const> draws`.